core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java

```diff
@@ -625,29 +625,33 @@ protected Object visitInPredicate(InPredicate node, Object context)
                 return set.contains(value);
             }
 
-            boolean hasUnresolvedValue = value instanceof Expression;
             boolean hasNullValue = false;
             boolean found = false;
-            List<Object> values = new ArrayList<>(valueList.getValues().size());
-            List<Type> types = new ArrayList<>(valueList.getValues().size());
+            ImmutableList.Builder<Object> values = ImmutableList.builderWithExpectedSize(valueList.getValues().size());
+            ImmutableList.Builder<Type> types = ImmutableList.builderWithExpectedSize(valueList.getValues().size());
 
             ResolvedFunction equalsOperator = metadata.resolveOperator(session, OperatorType.EQUAL, types(node.getValue(), valueList));
             for (Expression expression : valueList.getValues()) {
+                // Determine whether expression is null without processing it
+                if (expressionIsNull(expression)) {
```

Comment:
BTW this doesn't recognize some cases NULL value cases there were recognized before, so it looks like a regression

i.e. for these, the `process(expression, context)` would return `null`, but `expressionIsNull` returns `false`

- `CAST(JSON 'null' AS varchar)`
- `try_cast('abc' AS bigint)` ("safe" `Csat`)

Reply:
It's not a regression. It's a fast-path in the code that tries to detect "obvious" NULL expression cases. There is another code path below that actually processes IN values and checks for null results:

```
Object inValue = process(expression, context);
                if (inValue == null) {
                    hasNullValue = true;
                    continue;
                }
```

Reply:
I can add those cases to the test but they are working and passing

Reply:
Why do we want to have a fast-path for "obvious" NULLs?